pkg/connector/enterprise_applications.go

```diff
@@ -203,80 +210,71 @@ func (e *enterpriseApplicationsBuilder) Grants(ctx context.Context, resource *v2
-			ownersQuery.Set("$count", "true") // Required to prevent MS Graph from returning a 400
-		}
-
-		ownersURL := e.conn.buildBetaURL(path.Join("servicePrincipals", resource.Id.Resource, ownersStr), ownersQuery)
 		b.Push(pagination.PageState{
 			ResourceTypeID: ownersStr,
-			Token:          ownersURL,
 		})
-
-		appRoleAssignedToQuery := url.Values{}
-		appRoleAssignedToURL := e.conn.buildBetaURL(path.Join("servicePrincipals", resource.Id.Resource, "appRoleAssignedTo"), appRoleAssignedToQuery)
 		b.Push(pagination.PageState{
 			ResourceTypeID: assignmentStr,
-			Token:          appRoleAssignedToURL,
 		})
 	}
 
 	ps := b.Current()
 	switch ps.ResourceTypeID {
 	case assignmentStr:
-		resp := e.cacheGet(resource.Id.Resource).AppRolesAssignedTo
-		grants, err := slices.ConvertErr(resp, func(ara *appRoleAssignment) (*v2.Grant, error) {
-			var annos annotations.Annotations
-			rid := &v2.ResourceId{Resource: ara.PrincipalId}
-			switch ara.PrincipalType {
+		principalResp, err := e.cache.GetOrSet(principalId, func() (*client.ServicePrincipal, error) {
+			return e.client.ServicePrincipal(ctx, principalId)
+		})
+		if err != nil {
+			return nil, "", nil, err
+		}
+
+		resp := principalResp.AppRolesAssignedTo
+		grants, err := ConvertErr(resp, func(appRoleAssignment *client.AppRoleAssignment) (*v2.Grant, error) {
+			var options []grant.GrantOption
+
+			rid := &v2.ResourceId{Resource: appRoleAssignment.PrincipalId}
+			switch appRoleAssignment.PrincipalType {
 			case "User":
 				rid.ResourceType = userResourceType.Id
 			case "Group":
 				rid.ResourceType = groupResourceType.Id
-				annos.Update(&v2.GrantExpandable{
+
+				options = append(options, grant.WithAnnotation(&v2.GrantExpandable{
 					EntitlementIds: []string{
-						fmt.Sprintf("group:%s:members", ara.PrincipalId),
+						fmt.Sprintf("group:%s:members", appRoleAssignment.PrincipalId),
 					},
 					Shallow:         true,
 					ResourceTypeIds: []string{userResourceType.Id},
-				})
+				}))
 			case "ServicePrincipal":
 				// TODO: service principals can be managed identities, enterprise applications, or maybe something else entirely.
 				// We need to figure out the resource type instead of hard coding it to be a managed identity.
 				rid.ResourceType = managedIdentitylResourceType.Id
 				// rid.ResourceType = enterpriseApplicationResourceType.Id
+			default:
+				l.Error("baton-azure-infrastructure: unsupported PrincipalType type on app role assignment", zap.String("principal_type", appRoleAssignment.PrincipalType))
 			}
-			ur := &v2.Resource{Id: rid}
-			return &v2.Grant{
-				Id: ara.Id,
-				Entitlement: &v2.Entitlement{
-					Id: fmt.Sprintf("enterprise_application:%s:assignment:%s",
-						resource.Id.Resource,
-						ara.AppRoleId,
-					),
-					Resource: resource,
-				},
-				Principal:   ur,
-				Annotations: annos,
-			}, nil
+
+			return grant.NewGrant(
+				resource,
+				fmt.Sprintf("%s:assignment:%s",
```

Comment:
using `NewGrant` will construct the entitlement ID with the resource, so this will have resource ID in the ID twice. similar for `NewAssignmentEntitlement`/`NewPermissionEntitlement`